umap/umap_.py

```diff
@@ -1652,13 +1720,18 @@ def fit(self, X, y=None):
                 self.verbose,
             )
         else:
-            self._small_data = False
             # Standard case
+            self._small_data = False
+            # pynndescent doesn't accept callable functions for sparse data
+            if _HAVE_PYNNDESCENT and self._sparse_data and isinstance(self.metric, str):
```

Comment:
Since pynndescent is the "future" for all the NN computation that's where a lot of work is going, including efforts to generally speed things up for all the metrics that can be sped up. I am unlikely to port those improvements back to umap, so it would be good if we could pass the string version of the metric through to pynndescent if at all possible and let it choose the best option. That means you can simplify this condition to just if _HAVE_PYNNDESCENT:

Reply:
Sure 👍 Though I think it'd still have to be `if _HAVE_PYNNDESCENT and isinstance(self.metric, str)`, in order to account for a user passing a callable to `self.metric`, right?  
I chose to preferentially pass the callable version of the metric in case there were any named metrics in `umap` that didn't also exist in `pynndescent`.  Is that a condition worth worrying about?

Reply:
tried to do this in a reasonably safe way --
- when trying to import `pynndescent`, also import the names of distance functions defined in `pynndescent.distances` and `pynndescent.sparse`
- if `_HAVE_PYNNDESCENT`, prefer passing the string version of the metric as long as `pynndescent` also defines it (for the appropriate data type)

That should enable passing the string version more often as UMAP's additional metrics are ported over to `pynndescent`, while also remaining functional if new metrics are added to `umap`

Reply:
Yes -- that sounds like an excellent compromise. Thanks.